File: lib/ts/models/src/skos/SkosConcept.ts
import { mapTermToString } from "@paradicms/rdf";
import { rdf, skos } from "@paradicms/vocabularies";
import { Literal, NamedNode } from "@rdfjs/types";
import { Mixin } from "ts-mixer";
import { Memoize } from "typescript-memoize";
import { Concept } from "../Concept";
import { ResourceBackedModel } from "../ResourceBackedModel";
import { SomeImageThumbnailMixin } from "../SomeImageThumbnailMixin";
import { Text } from "../Text";
import { FoafImagesMixin } from "../foaf/FoafImagesMixin";
import { mapTermToText } from "../mapTermToText";

export class SkosConcept
  extends Mixin(ResourceBackedModel, FoafImagesMixin, SomeImageThumbnailMixin)
  implements Concept {
  @Memoize()
  get altLabels(): readonly Literal[] {
    return this.filterAndMapObjects(skos.altLabel, term =>
      term.termType === "Literal" ? (term as Literal) : null
    );
  }

  @Memoize()
  get definition(): Text | null {
    return this.findAndMapObject(skos.definition, term =>
      mapTermToText(this, term)
    );
  }

  get description(): Text | null {
    return this.definition;
  }

  get label(): string {
    if (this.prefLabel) {
      return this.prefLabel;
    }
    for (const altLabel of this.altLabels) {
      return altLabel.value;
    }
    return this.value.value;
  }

  @Memoize()
  get prefLabel(): string | null {
    return this.findAndMapObject(skos.prefLabel, mapTermToString);
  }

  override preMemoize(): void {
    super.preMemoize();
    this.preMemoizeImages();
    this.altLabels;
    this.definition;
    this.prefLabel;
    this.value;
  }

  @Memoize()
  get value(): Literal | NamedNode {
    return (
      this.findAndMapObject(rdf.value, term => {
        switch (term.termType) {
          case "Literal":
          case "NamedNode":
            return term as Literal | NamedNode;
          default:
            return null;
        }
      }) ?? this.identifier
    );
  }
}

File: lib/py/etl/paradicms_etl/models/skos/skos_concept.py
from __future__ import annotations

from typing import TYPE_CHECKING

from rdflib import SKOS, Graph
from rdflib.namespace import RDF
from rdflib.resource import Resource
from rdflib.term import Literal, Node, URIRef

from paradicms_etl.models.concept import Concept
from paradicms_etl.models.foaf.foaf_images_mixin import FoafImagesMixin
from paradicms_etl.models.resource_backed_model import ResourceBackedModel
from paradicms_etl.utils.safe_dict_update import safe_dict_update

if TYPE_CHECKING:
    from paradicms_etl.models.text import Text


class SkosConcept(ResourceBackedModel, FoafImagesMixin, Concept):
    class Builder(ResourceBackedModel.Builder, FoafImagesMixin.Builder):
        def add_alt_label(self, alt_label: str | Literal) -> SkosConcept.Builder:
            self.add(SKOS.altLabel, alt_label)
            return self

        def add_type_uri(self, type_uri: URIRef) -> SkosConcept.Builder:
            self.add(RDF.type, type_uri)
            return self

        def build(self) -> SkosConcept:
            return SkosConcept(self._resource)

        def set_definition(self, definition: str | Text) -> SkosConcept.Builder:
            self.set(SKOS.definition, definition)
            return self

        def set_value(self, value: Node) -> SkosConcept.Builder:
            self.set(RDF.value, value)
            return self

    def __init__(self, resource: Resource):
        resource.add(RDF.type, SKOS.Concept)
        ResourceBackedModel.__init__(self, resource)
        self.label

    @classmethod
    def builder(cls, *, pref_label: str, uri: URIRef) -> Builder:
        builder = cls.Builder(Graph().resource(uri))
        builder.set(SKOS.prefLabel, pref_label)
        return builder

    @classmethod
    def json_ld_context(cls):
        return safe_dict_update(
            ResourceBackedModel.json_ld_context(),
            FoafImagesMixin.json_ld_context(),
            {
                "altLabel": {"@id": str(SKOS.altLabel)},
                "definition": {"@id": str(SKOS.definition)},
                "prefLabel": {"@id": str(SKOS.prefLabel)},
                "value": {"@id": str(RDF.value)},
            },
        )

    @property
    def label(self) -> str:
        return self._required_label

    @classmethod
    def label_property_uri(cls) -> URIRef:
        return SKOS.prefLabel

    @classmethod
    def rdf_type_uri(cls) -> URIRef:
        return SKOS.Concept

    @property
    def type_uris(self) -> tuple[URIRef, ...]:
        return tuple(self._values(RDF.type, self._map_term_to_uri))

    @property
    def value(self) -> Node:
        value = self.resource.value(RDF.value)
        if value is None:
            return self.uri
        if isinstance(value, Resource):
            return value.identifier
        assert isinstance(value, Node)
        return value
